File: ring-android/libringclient/src/main/java/cx/ring/mvp/RootPresenter.java
/*
 *  Copyright (C) 2016 Savoir-faire Linux Inc.
 *
 *  Author: Thibault Wittemberg <thibault.wittemberg@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
package cx.ring.mvp;

import java.lang.ref.WeakReference;

public class RootPresenter<T> {

    private WeakReference<T> mView;

    public void bindView(T view) {
        mView = new WeakReference<>(view);
    }

    public void unbindView() {
        if (mView != null) {
            mView.clear();
        }

        mView = null;
    }

    public T getView() {
        if (mView != null) {
            return mView.get();
        }

        return null;
    }

}

File: jami-android/libjamiclient/src/main/kotlin/net/jami/mvp/RootPresenter.kt
/*
 *  Copyright (C) 2004-2025 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package net.jami.mvp

import io.reactivex.rxjava3.disposables.CompositeDisposable
import java.lang.ref.WeakReference

abstract class RootPresenter<T> {
    protected val mCompositeDisposable = CompositeDisposable()
    private var mView: WeakReference<T>? = null

    open fun bindView(view: T) {
        mView = WeakReference(view)
    }

    open fun unbindView() {
        mView?.let { view ->
            view.clear()
            mView = null
        }
        mCompositeDisposable.clear()
    }

    open fun onDestroy() {
        mCompositeDisposable.dispose()
    }

    val view: T?
        get() = mView?.get()
}